internal/provider/datasources/deployment.go

```diff
@@ -0,0 +1,405 @@
+	if !model.ID.IsNull() {
+		var deploymentID uuid.UUID
+		deploymentID, err = uuid.Parse(model.ID.ValueString())
+		if err != nil {
+			resp.Diagnostics.AddAttributeError(
+				path.Root("id"),
+				"Error parsing Deployment ID",
+				fmt.Sprintf("Could not parse deployment ID to UUID, unexpected error: %s", err.Error()),
+			)
+
+			return
+		}
+
+		operation = "get"
+		deployment, err = client.Get(ctx, deploymentID)
+	} else if !model.Name.IsNull() {
+		var deployments []*api.Deployment
+		operation = "list"
+		deployments, err = client.List(ctx, []string{model.Name.ValueString()})
+
+		if len(deployments) != 1 {
+			resp.Diagnostics.AddError(
+				"Could not find Deployment",
+				fmt.Sprintf("Could not find Deployment with name %s", model.Name.ValueString()),
+			)
+
+			return
+		}
+
+		deployment = deployments[0]
+	}
+
+	if err != nil {
+		resp.Diagnostics.Append(helpers.ResourceClientErrorDiagnostic("Deployment", operation, err))
+
+		return
+	}
+
+	resp.Diagnostics.Append(copyDeploymentToModel(ctx, deployment, &model)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	parametersByteSlice, err := json.Marshal(deployment.Parameters)
+	if err != nil {
+		resp.Diagnostics.Append(helpers.SerializeDataErrorDiagnostic("parameters", "Deployment parameters", err))
+	}
+	model.Parameters = jsontypes.NewNormalizedValue(string(parametersByteSlice))
+
+	jobVariablesByteSlice, err := json.Marshal(deployment.JobVariables)
+	if err != nil {
+		resp.Diagnostics.Append(helpers.SerializeDataErrorDiagnostic("job_variables", "Deployment job variables", err))
+	}
+	model.JobVariables = jsontypes.NewNormalizedValue(string(jobVariablesByteSlice))
+
+	parameterOpenAPISchemaByteSlice, err := json.Marshal(deployment.ParameterOpenAPISchema)
+	if err != nil {
+		resp.Diagnostics.Append(helpers.SerializeDataErrorDiagnostic("parameter_openapi_schema", "Deployment parameter OpenAPI schema", err))
+	}
+	model.ParameterOpenAPISchema = jsontypes.NewNormalizedValue(string(parameterOpenAPISchemaByteSlice))
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &model)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Configure initializes runtime state for the data source.
+func (d *deploymentDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(api.PrefectClient)
+	if !ok {
+		resp.Diagnostics.Append(helpers.ConfigureTypeErrorDiagnostic("data source", req.ProviderData))
+
+		return
+	}
+
+	d.client = client
+}
+
+// copyDeploymentToModel leverages the function by the same name from the resources package to avoid repeating
+// the logic. Because DeploymentResourceModel embeds the resource.DeploymentResourceModel type, we can cast
+// it to the compatible type before calling the referenced function.
+func copyDeploymentToModel(ctx context.Context, deployment *api.Deployment, model *DeploymentDataSourceModel) diag.Diagnostics {
```

Comment:
This accompanies the embedded struct for the data source model. It's a bit clunky to translate between the two types, but imo it beats repeating all of the logic in CopyDeploymentToModel. open to feedback though

Reply:
well this is pretty smart